pkg/installmanager/installmanager_test.go

```diff
@@ -865,6 +866,205 @@ spec:
+  template:
+    metadata:
+      labels:
+        "machine.openshift.io/cluster-api-cluster": "test-cluster"
+        "machine.openshift.io/cluster-api-machine-role": "edge"
+		"machine.openshift.io/cluster-api-machine-type": "edge"`,
+			expectModified:  false,
+			expectErr:       true,
+			machinePoolType: "edge",
+		},
+		{
+			name: "MachineSet spec.template.metadata section does not exist", // FIXME do we need this?
+			manifestYAML: `---
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineSet
+spec:
+  template: {}`,
+			expectModified:  false,
+			expectErr:       true,
+			machinePoolType: "edge",
+		},
+		{
+			name: "metadata.labels section already exists and is not overwritten",
+			manifestYAML: `---
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineSet
+metadata:
+  labels:
+    "potato": "salad"
+spec:
+  template:
+    metadata:
+      labels:
+        "machine.openshift.io/cluster-api-cluster": "test-cluster"
+        "machine.openshift.io/cluster-api-machine-role": "edge"
+        "machine.openshift.io/cluster-api-machine-type": "edge"
+        "machine.openshift.io/cluster-api-machineset": "test-cluster-edge-us-east-1a"`,
+			expectModified:     true,
+			machinePoolType:    "edge",
+			testExistingLabels: map[string]string{"potato": "salad"},
+		},
+	}
+	for _, tc := range cases {
+
+		t.Run(tc.name, func(t *testing.T) {
+
+			pool := &hivev1.MachinePool{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "testpool",
+					Namespace: "testnamespace",
+				},
+			}
+
+			pool.Spec.Name = tc.machinePoolType
+
+			mapPoolsByType := map[string]string{tc.machinePoolType: pool.Name}
+
+			logger := log.WithFields(log.Fields{"machinePool": pool.Name})
+
+			modifiedBytes, err := patchLabelMachineSetManifest([]byte(tc.manifestYAML), mapPoolsByType, logger)
+			if tc.expectModified {
+				assert.NotNil(t, modifiedBytes, "expected manifest to be modified")
+			} else {
+				assert.Nil(t, modifiedBytes, "expected manifest to not be modified")
+			}
+			if tc.expectErr {
+				assert.Error(t, err, "expected error patching worker machineset manifests")
+			} else {
+				assert.NoError(t, err, "unexpected error patching worker machineset manifests")
+			}
+
+			scheme := scheme.GetScheme()
+			codecFactory := serializer.NewCodecFactory(scheme)
+			decoder := codecFactory.UniversalDecoder(machineapi.SchemeGroupVersion)
+
+			if tc.expectModified {
+				machineSetObj, _, err := decoder.Decode(*modifiedBytes, nil, nil)
+				assert.NoError(t, err, "expected to be able to decode MachineSet yaml")
+				machineSet, _ := machineSetObj.(*machineapi.MachineSet)
+
+				assert.Equal(t, pool.Spec.Name, machineSet.ObjectMeta.Labels[machinePoolNameLabel], "expected label %s to be %s", machinePoolNameLabel, pool.Spec.Name)
+
+				if tc.testExistingLabels != nil {
+					for k, v := range tc.testExistingLabels {
+						assert.Equal(t, v, machineSet.ObjectMeta.Labels[k], "expected label %s to be %s", k, v)
+					}
+				}
+			}
+
+		})
+	}
+}
```

Comment:
Good tests. I feel like if we were going past 80/20 we would also
- run this suite with multiple MPs and/or msets present, e.g. to make sure there isn't bleed-over
- cover the WalkDir bit

but this works for now.